File: practice/cpp/COURSERA/competProgr/longestIncSubS/sequence.py
def lis(A):
    T = [None] * len(A)
    prev = [None] * len(A)

    for i in range(len(A)):
        T[i] = 1
        prev[i] = -1
        for j in range(i):
            if A[j] < A[i] and T[i] < T[j] + 1:
                T[i] = T[j] + 1
                prev[i] = j
                
    last = 0
    for i in range(1, len(A)):
        if T[i] > T[last]:
            last = i

    lis= []
    current = last
    while current >= 0:
        lis.append(current)
        current = prev[current]
    lis.reverse()
    return [A[i] for i in lis]

print(lis([7, 2, 1, 3, 8, 4, 9, 1, 2, 6, 5, 9, 3, 8, 1]))

File: practice/cpp/COURSERA/competProgr/longestIncSubS/sequence.cpp
#include<iostream>
#include<vector>
#define printVect(vect,dataStruct) for(vector<dataStruct> :: iterator it = vect.begin(); it !=vect.end();it++){cout << *it << " ";}
using namespace std;
vector<int> list(vector<int> A,vector<int> B){
    vector<int> T,prev;
    for(int i = 0; i < A.size(); i++){
            T.push_back(1) ;
            prev.push_back(-1) ;
        for(int j = 0; j < i; j++){
            if(A[j] < A[i] && T[i] < T[j] + 1){
                T[i] = T[j] + 1;
                prev[i] = j;
            }
        }
    }

    int last = 0;
    for(int i = 0; i < A.size(); i++){
        if(T[i] > T[last]){
            last = i;
        }
    }
    int current  = last;
    while(current >= 0){
        B.insert(B.begin(),current);
        current = prev[current];
    }
    return B;
}

int main(){
    vector<int> A,B;
    A = {7, 2, 1, 3, 8, 4, 9, 1, 2, 6, 5, 9, 3, 8, 1};
    B = list(A,B);
    // printVect(B,int);
    cout << "{" ;
    for(int i = 0; i < B.size(); i++){
        if(i != B.size()-1){
             cout << A[B[i]] << "," ;
        }else{
            cout << A[B[i]];
        }
       
    }
    cout << "}" << endl;
    return 0;
}